packages/integrations/components/pico/utils.js

```diff
@@ -0,0 +1,33 @@
+export default function mountPicoNodes(pageInfo) {
+  let nodesMounted = false;
+
+  // Mount the Signal and Rule nodes to be targeted by the Pico embed.
+  const documentBody = document.getElementsByTagName('body')[0];
```

Comment:
Are the elements added here defined by the Pico scripts? If so, is there a scenario we're correcting for here where they're not added by Pico's own embed? If not, is there a rationale for storing these values in document elements instead of the Redux store?

Reply:
@xmattus Great question! From my understanding of the [Pico documentation](https://help.trypico.com/en/articles/2906319-using-pico-signals) (which there isn't much of any it isn't really centered on a developer's perspective), they do not provide any container elements that are have the target `PicoSignal` class name. Pico uses this element as a sort of container for their widget to read and store user data based on the actions the user takes within their embed.

This element is being defined here outside of the root context of the Irving application so that it is available to Pico at all times. It gets injected on an initial server-load and is available through every subsequent client-side route change because it lives outside of Irving's root node. I had to go this route because Pico fires two events on every `visit`, `pico.init` and `pico.load`. If the target elements (both the signal and rule nodes) don't exist in the DOM by the time those actions are fired, Pico won't be able to find them and execute the appropriate behavior. It was possible to mount these elements within the context of the `withPico` HOC wrapper, but that ended up getting pretty messy and I had to resort to listening for `LOCATION_CHANGED` actions and delaying firing the `visit` event until I knew for a fact that the component had rendered.

As far as storing these values in the Redux store goes, it would definitely be doable, but I don't think it's necessary in this context. Since Pico targets any DOM element with the `PicoSignal` class and adds the user information as data attributes to that node automatically I don't really see a need for that information to exist in multiple places at once.

Reply:
Ah, ok. That makes sense, and I recall a bit of this from when I did the Pico implementation for the JFC 1.0 (née Civil First Fleet) sites.